src/Features/Core/Portable/GenerateConstructorFromMembers/AbstractGenerateConstructorFromMembersCodeRefactoringProvider.cs

```diff
@@ -47,60 +49,149 @@ protected AbstractGenerateConstructorFromMembersCodeRefactoringProvider() : this
+            await ComputeRefactoringsAsync(
+                priorDocument,
+                priorSelection,
+                (singleAction, applicableToSpan) => actions.Add(singleAction),
+                (multipleActions) => actions.AddRange(multipleActions),
+                cancellationToken).ConfigureAwait(false);
+
+            if (actions.IsEmpty())
+            {
+                return ImmutableArray<IntentProcessorResult>.Empty;
+            }
+
+            // The refactorings returned will be in the following order (if available)
+            // FieldDelegatingCodeAction, ConstructorDelegatingCodeAction, GenerateConstructorWithDialogCodeAction
+            using var resultsBuilder = ArrayBuilder<IntentProcessorResult>.GetInstance(out var results);
+            foreach (var action in actions)
+            {
+                var intentResult = await GetIntentProcessorResultAsync(action, cancellationToken).ConfigureAwait(false);
+                results.AddIfNotNull(intentResult);
+            }
+
+            return results.ToImmutable();
+
+            static async Task<IntentProcessorResult?> GetIntentProcessorResultAsync(CodeAction codeAction, CancellationToken cancellationToken)
+            {
+                var title = codeAction.Title;
+                var operations = await GetCodeActionOperationsAsync(codeAction, cancellationToken).ConfigureAwait(false);
+
+                // Generate ctor will only return an ApplyChangesOperation or potentially document navigation actions.
+                // We can only return edits, so we only care about the ApplyChangesOperation.
+                var applyChangesOperation = operations.OfType<ApplyChangesOperation>().SingleOrDefault();
+                if (applyChangesOperation == null)
+                {
+                    return null;
+                }
+
+                var type = typeof(CodeAction);
+                return new IntentProcessorResult(applyChangesOperation.ChangedSolution, title, type.Name);
+            }
+
+            static async Task<ImmutableArray<CodeActionOperation>> GetCodeActionOperationsAsync(
+                CodeAction action,
+                CancellationToken cancellationToken)
+            {
+                if (action is CodeActionWithOptions)
+                {
+                    // Generate ctor only returns GenerateConstructorWithDialogCodeAction for CodeActionWithOptions
+                    var dialogAction = (GenerateConstructorWithDialogCodeAction)action;
+
+                    // Usually applying this code action pops up a dialog allowing the user to choose which options.
+                    // We can't do that here, so instead we just take the defaults until we have more intent data.
+                    var options = new PickMembersResult(dialogAction.ViableMembers, dialogAction.PickMembersOptions);
+                    var operations = await dialogAction.GetOperationsAsync(options: options, cancellationToken).ConfigureAwait(false);
```

Comment:
i generally hate this as it looks like it is popping up UI here :)  is there like an extracted static helper we could call instead that just has the logic?

Reply:
Hmm I think the CodeActionWithOptions contract is fairly clear - the options are only determined when GetOptions is called.  If the API teakes in the options, the options have already been determined.  But I will extract this out in a followup where I do some more cleanup on the implementation here.